cert/aziot-certd/src/lib.rs

```diff
@@ -393,18 +393,58 @@ async fn create_cert_inner<'a>(
             }
         }
 
-        let (subject_name, expiry) = stack.get(0).map_or((subject_name, expiry), |x509| {
-            let issuer_expiry = x509.not_after();
+        let (subject_name, expiry, issuer_ref) =
+            stack.get(0).map_or((subject_name, expiry, None), |x509| {
+                let issuer_expiry = x509.not_after();
+
+                (
+                    x509.subject_name(),
+                    if expiry < issuer_expiry {
+                        expiry
+                    } else {
+                        issuer_expiry
+                    },
+                    Some(x509.as_ref()),
+                )
+            });
+
+        let subj_key_id = openssl::x509::extension::SubjectKeyIdentifier::new();
+        let context = builder.x509v3_context(issuer_ref, None);
+        let subj_key_id = subj_key_id.build(&context)?;
+        builder.append_extension(subj_key_id)?;
+
+        let mut auth_key_id = openssl::x509::extension::AuthorityKeyIdentifier::new();
+        auth_key_id.keyid(true);
+
+        let auth_key_id = if issuer_ref.is_none() {
+            // Issuer is None: this is a self-signed CA cert. Despite the openssl documentation
+            // saying that the issuer context should be None for self-signed certificates, this
+            // fails at runtime, as openssl is not able to determine the key ID when given None.
+            //
+            // Workaround: generate another cert that bears the same key as the cert being issued
+            // and use that as the issuer.
+            let mut issuer_builder = X509::builder()?;
+            issuer_builder.set_version(2)?;
+            issuer_builder.set_subject_name(subject_name)?;
+            issuer_builder.set_pubkey(pubkey)?;
+            issuer_builder.set_not_before(&*Asn1Time::days_from_now(0)?)?;
+            issuer_builder.set_not_after(expiry)?;
+
+            let issuer_key_id = openssl::x509::extension::SubjectKeyIdentifier::new();
+            let issuer_context = issuer_builder.x509v3_context(None, None);
+            let issuer_key_id = issuer_key_id.build(&issuer_context)?;
+            issuer_builder.append_extension(issuer_key_id)?;
+
+            let issuer = issuer_builder.build();
+
+            let context = builder.x509v3_context(Some(&issuer), None);
+            auth_key_id.build(&context)?
+        } else {
+            let context = builder.x509v3_context(issuer_ref, None);
+            auth_key_id.build(&context)?
```

Comment:
What does openssl do if the issuer doesn't have a SKID? Does it no-op the AKID extension or does it fail? If it fails, we should not fail the cert creation and instead just skip the AKID extension.

Reply:
Fixed.